File: permutationIndex.py
class Solution:
    """
    @param A: An array of integers
    @return: A long integer
    """
    def permutationIndex(self, A):
        # write your code here
        ret =1
        factor =1
        #start from the end of the array
        for i in range(len(A)-1, -1, -1):
            rank = 0
            #find the number of elment that smaller than i after i
            for j in range(i+1, len(A)):
                if A[j] < A[i] :
                    rank +=1
            #cantor extention
            ret += factor*rank
            factor *= len(A)-i
            
        return ret

File: permutation/permutationIndex.cc
class Solution {
public:
    /**
     * @param A: An array of integers
     * @return: A long integer
     */
    long long permutationIndex(vector<int> &A) {
        // write your code here
        
        long long ret= 1;
        int n= A.size();
        long long factor= 1;
        for(int i=n-1; i>=0; i--) {
            int rank =0;
            for(int j= i+1; j<n; j++)
                if(A[j] < A[i])
                    rank++;
            ret += rank*factor;
            factor *= n-i;
        }
        
        return ret;
    }
};
